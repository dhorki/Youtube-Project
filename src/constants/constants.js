export const styles = {
  colors: {
    light: {
      appIcon: '#ff0200',
      background: '#F9F9F9',
      barBackground: '#ffffff',
      primary: '#030303',
      secondary: '#606060',
      linkActive: '#ff0200',
      linkHoverBackground: '#ececec',
      inputBackground: '#ffffff',
      inputBorder: '#cccccc',
      searchButtonBackground: '#f8f8f8',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
    dark: {
      appIcon: '#ff0200',
      background: '#181818',
      barBackground: '#212121',
      primary: '#ffffff',
      secondary: '#aaaaaa',
      linkActive: '#ff0200',
      linkHoverBackground: '#303030',
      inputBackground: '#121212',
      inputBorder: '#303030',
      searchButtonBackground: '#323232',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
  },
  props: {
    transitionTime: '0.4s',
    sidebarWidth: '150px',
  },
};

export const screenSizes = {
  xsmall: '675px',
  small: '830px',
  mid: '1200px',
};

export const mainViewTypes = {
  searchGallery: 'searchGallery',
  favoritesGallery: 'favoritesGallery',
  videoDetails: 'videoDetails',
};

export const environmentActions = {
  setTheme: 'setTheme',
  setSidebarShow: 'setSidebarShow',
  setSearchQuery: 'setSearchQuery',
  setModalShow: 'setModalShow',
  setUser: 'setUser',
};

export const modalTypes = {
  login: 'login',
  register: 'register',
  resetPassword: 'reset-password',
  none: 'none',
};

export const resetSearchQueryAction = {
  type: environmentActions.setSearchQuery,
  payload: {
    q: '',
  },
};

export const hideModalAction = {
  type: environmentActions.setModalShow,
  payload: {
    modalShow: modalTypes.none,
  },
};
