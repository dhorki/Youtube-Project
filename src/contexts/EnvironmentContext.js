import { createContext } from 'react';

export const EnvironmentContext = createContext();
