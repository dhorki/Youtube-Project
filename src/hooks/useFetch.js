import { useEffect, useRef, useState } from 'react';

const cache = { current: {} };

export const useFetch = (url, cacheIt = true) => {
  const _isMounted = useRef(true); // Initial value _isMounted = true
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    let controller = new AbortController();
    // console.log(cache);

    const fetchData = async () => {
      setLoading(true);
      const cachedResult = cache.current[url];
      if (cachedResult && !cachedResult.error) {
        const result = cache.current[url];
        // console.log('fetching from cache ' + url);
        setData(result);
      } else {
        try {
          // console.log('fetching');
          const response = await fetch(url, { signal: controller.signal });
          const result = await response.json();
          if (result.error) {
            setError(result.error);
          }
          if (cacheIt) {
            cache.current[url] = { ...result, fromCache: true }; // set response in cache;
          }
          setData(result);
        } catch (e) {
          // console.log(e);
          if (_isMounted.current) {
            setError(e);
          }
        }
      } //else
      if (_isMounted.current) {
        setLoading(false);
      }
    };
    fetchData();

    return () => {
      _isMounted.current = false;
      controller.abort();
    };
  }, [url, cacheIt]);

  return {
    data,
    loading,
    error,
  };
};
