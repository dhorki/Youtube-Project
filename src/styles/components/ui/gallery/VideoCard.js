import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { styles } from '../../../../constants/constants';

export const StyledVideoCard = styled(Link)`
  display: flex;
  flex-direction: column;
  width: 360px;

  margin: 10px;
  margin-bottom: 40px;
  background-color: ${({ theme }) => theme.background};
  color: ${({ theme }) => theme.primary};
  text-decoration: none;

  transition: background-color ${styles.props.transitionTime} ease;

  > img {
    height: 200px;
  }

  > .videocard-title {
    margin-top: 10px;
    color: ${({ theme }) => theme.primary};
    font-size: 18px;
    font-weight: bold;
  }

  > .videocard-channel-title {
    margin-top: 10px;
    color: ${({ theme }) => theme.secondary};
    font-size: 14px;
  }

  > .videocard-description {
    margin-top: 10px;
    color: ${({ theme }) => theme.secondary};
    font-size: 14px;
  }
`;
