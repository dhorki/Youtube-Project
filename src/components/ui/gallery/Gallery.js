import React from 'react';
import { VideoCard } from './VideoCard';
import { StyledGallery } from '../../../styles/components/ui/gallery/Gallery';
import { useFetch } from '../../../hooks/useFetch';
import { useLocation } from 'react-router-dom';
import { mainViewTypes } from '../../../constants/constants';

export const Gallery = () => {
  const location = useLocation();

  const galleryType =
    location.pathname === '/' ? mainViewTypes.searchGallery : mainViewTypes.videoGallery;
  const queryString = require('query-string');
  const { q = '' } = queryString.parse(location.search);

  const searchUrl =
    'https://www.googleapis.com/youtube/v3/search?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=snippet&' +
    `q=${q}&` +
    'type=video&' +
    'order=rating&' +
    'maxResults=24';

  const favoritesUrl =
    'https://www.googleapis.com/youtube/v3/videos?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=id,snippet,player&' +
    'id=id1,id2,id3&' +
    'maxResults=24';

  let url = '';
  if (galleryType === mainViewTypes.searchGallery) {
    url = searchUrl;
  } else {
    url = favoritesUrl;
  }

  const { data, loading, error } = useFetch(url);
  // const { data, loading, error } = { data: [], loading: true, error: null };

  let filteredItems = [];

  if (!loading && !error) {
    filteredItems = data.items?.filter((item) => item.id.kind.includes('youtube#video'));
  }

  const getNoResultsMessage = () => {
    if (galleryType === mainViewTypes.searchGallery) {
      return q ? <p>No results found for your search.</p> : null;
    } else {
      return <p>You have no favorites yet.</p>;
    }
  };

  return (
    <StyledGallery>
      {loading ? (
        <div className="loading">
          <img
            src="https://i.pinimg.com/originals/d4/2b/b2/d42bb2f00e1042c990b25f90ada9b0ab.gif"
            alt="loading"
          />
        </div>
      ) : error ? (
        <p>Error: {error?.message}</p>
      ) : !filteredItems || filteredItems.length === 0 ? (
        getNoResultsMessage()
      ) : (
        filteredItems.map((video) => <VideoCard key={video.id.videoId} video={video} />)
      )}
    </StyledGallery>
  );
};
