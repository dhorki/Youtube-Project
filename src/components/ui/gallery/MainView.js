import React, { useContext } from 'react';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';
import { useTheme } from '../../../hooks/useTheme';
import { Gallery } from './Gallery';
import { StyledMainView } from '../../../styles/components/ui/gallery/MainView';

export const MainView = () => {
  const { getTheme } = useTheme();
  const theme = getTheme();

  const { sidebarShow } = useContext(EnvironmentContext);

  return (
    <StyledMainView theme={theme} sidebarShow={sidebarShow}>
      <div className="main-container">
        <Gallery />
      </div>
    </StyledMainView>
  );
};
