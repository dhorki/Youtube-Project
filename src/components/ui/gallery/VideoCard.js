import React, { useContext } from 'react';
import { timeSince } from '../../../helpers/formatters';
import { decodeHtml } from '../../../helpers/decodeHtml';
import { StyledVideoCard } from '../../../styles/components/ui/gallery/VideoCard';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';
import PropTypes from 'prop-types';

export const VideoCard = ({ video }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const { id, snippet } = video;
  const { videoId } = id;
  const { title, channelTitle, description, thumbnails, publishTime } = snippet;
  const { url } = thumbnails.medium;
  const d = new Date(publishTime);

  return (
    <StyledVideoCard
      className="animate__animated animate__fadeIn"
      theme={theme}
      to={`/${videoId}`}
    >
      <img src={url} alt="video" />
      <div className="videocard-title">{decodeHtml(title)}</div>
      <div className="videocard-channel-title">
        {decodeHtml(channelTitle)} - {timeSince(d)} ago.
      </div>
      <div className="videocard-description">{decodeHtml(description)}</div>
    </StyledVideoCard>
  );
};

VideoCard.propTypes = {
  video: PropTypes.object.isRequired,
};
