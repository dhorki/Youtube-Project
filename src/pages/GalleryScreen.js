import React, { useContext } from 'react';
import { VideoCard } from '../components/ui/gallery/VideoCard';
import { StyledGallery } from '../styles/pages/Gallery';
import { useFetch } from '../hooks/useFetch';
import { useLocation } from 'react-router-dom';
import { mainViewTypes } from '../constants/constants';
import { LoadingAnimation } from '../components/ui/loading/LoadingAnimation';
import { ErrorText } from '../components/ui/error/ErrorText';
import { EnvironmentContext } from '../contexts/EnvironmentContext';

export const GalleryScreen = () => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const location = useLocation();

  const galleryType =
    location.pathname === '/' ? mainViewTypes.searchGallery : mainViewTypes.videoGallery;
  const queryString = require('query-string');
  const { q = '' } = queryString.parse(location.search);

  const searchUrl =
    'https://www.googleapis.com/youtube/v3/search?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=snippet&' +
    `q=${q}&` +
    'type=video&' +
    `order=${q.length === 0 ? 'date' : 'relevance'}&` +
    'regionCode=MX&' +
    'relevanceLanguage=es&' +
    'maxResults=24';

  const favoritesUrl =
    'https://www.googleapis.com/youtube/v3/videos?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=id,snippet&' +
    'id=id1,id2,id3&' +
    'maxResults=24';

  let url = '';
  if (galleryType === mainViewTypes.searchGallery) {
    url = searchUrl;
  } else {
    url = favoritesUrl;
  }

  const { data, loading, error } = useFetch(url, false);
  // const { data, loading, error } = { data: [], loading: true, error: null };

  let filteredItems = [];

  if (!loading && !error) {
    filteredItems = data.items?.filter((item) => item.id.kind.includes('youtube#video'));
  }

  const getNoResultsMessage = () => {
    let message = '';
    if (galleryType === mainViewTypes.searchGallery) {
      q && (message = 'No results found for your search.');
    } else {
      message = 'You have no favorites yet';
    }

    return message ? (
      <p className="animate__animated animate__fadeIn">{message}</p>
    ) : null;
  };

  return (
    <StyledGallery theme={theme}>
      {loading ? (
        <LoadingAnimation />
      ) : error ? (
        <ErrorText error={error} />
      ) : !filteredItems || filteredItems.length === 0 ? (
        getNoResultsMessage()
      ) : (
        filteredItems.map((video) => <VideoCard key={video.id.videoId} video={video} />)
      )}
    </StyledGallery>
  );
};
