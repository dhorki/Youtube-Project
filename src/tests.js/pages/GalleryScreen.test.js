import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { styles } from '../../constants/constants';
import { EnvironmentContext } from '../../contexts/EnvironmentContext';
import { GalleryScreen } from '../../pages/GalleryScreen';
import { useFetch } from '../../hooks/useFetch';
import { MemoryRouter } from 'react-router-dom';
import { VideoCard } from '../../components/ui/gallery/VideoCard';
jest.mock('../../hooks/useFetch');
jest.mock('../../components/ui/gallery/VideoCard');

describe('GalleryScreen', () => {
  const environment = {
    theme: styles.colors.light,
    sidebarShow: false,
    q: '',
  };

  it('should render correctly while loading', () => {
    const videoId = 'videoId';

    const mockLocation = {
      pathname: '/',
      search: '?q=random',
    };

    useFetch.mockReturnValue({
      data: {},
      loading: true,
      error: null,
    });

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <GalleryScreen />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with success fetch as search', () => {
    const mockLocation = {
      pathname: '/',
      search: '?q=random',
    };

    useFetch.mockReturnValue({
      data: {
        items: [
          {
            id: {
              videoId: 'videoId1',
              kind: 'youtube#video',
            },
          },
          {
            id: {
              videoId: 'videoId2',
              kind: 'youtube#video',
            },
          },
          {
            id: {
              videoId: 'videoId3',
              kind: 'youtube#video',
            },
          },
        ],
      },
      loading: false,
      error: null,
    });

    VideoCard.mockImplementation(({ video }) => <div>VideoCard {video.id.videoId}</div>);

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <GalleryScreen />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with success fetch as favorites', () => {
    const mockLocation = {
      pathname: '/favorites',
      search: '',
    };

    useFetch.mockReturnValue({
      data: {
        items: [
          {
            id: {
              videoId: 'videoId1',
              kind: 'youtube#video',
            },
          },
          {
            id: {
              videoId: 'videoId2',
              kind: 'youtube#video',
            },
          },
          {
            id: {
              videoId: 'videoId3',
              kind: 'youtube#video',
            },
          },
        ],
      },
      loading: false,
      error: null,
    });

    VideoCard.mockImplementation(({ video }) => <div>VideoCard {video.id.videoId}</div>);

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <GalleryScreen />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with success fetch as search with no results', () => {
    const mockLocation = {
      pathname: '/',
      search: '?q=random',
    };

    useFetch.mockReturnValue({
      data: {
        items: [],
      },
      loading: false,
      error: null,
    });

    VideoCard.mockImplementation(({ video }) => <div>VideoCard {video.id.videoId}</div>);

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <GalleryScreen />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with success fetch as favorites with no results', () => {
    const mockLocation = {
      pathname: '/favorites',
      search: '',
    };

    useFetch.mockReturnValue({
      data: {
        items: [],
      },
      loading: false,
      error: null,
    });

    VideoCard.mockImplementation(({ video }) => <div>VideoCard {video.id.videoId}</div>);

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <GalleryScreen />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with fail fetch', () => {
    const mockLocation = {
      pathname: '/favorites',
      search: '?q=',
    };

    useFetch.mockReturnValue({
      data: {},
      loading: false,
      error: { error: 'error' },
    });

    VideoCard.mockImplementation(({ video }) => <div>VideoCard {video.id.videoId}</div>);

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <MemoryRouter initialEntries={[mockLocation]}>
          <GalleryScreen />
        </MemoryRouter>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
