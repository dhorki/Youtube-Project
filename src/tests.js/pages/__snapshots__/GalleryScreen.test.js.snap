// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryScreen should render correctly while loading 1`] = `
<div
  className="sc-bczRLJ ljydhb"
>
  <div
    className="sc-gsnTZi gEmEgX"
  >
    <img
      alt="loading"
      src="https://i.pinimg.com/originals/d4/2b/b2/d42bb2f00e1042c990b25f90ada9b0ab.gif"
    />
  </div>
</div>
`;

exports[`GalleryScreen should render correctly with fail fetch 1`] = `
<div
  className="sc-bczRLJ ljydhb"
>
  <p
    className="animate__animated animate__fadeIn"
  >
    Error: 
    undefined
  </p>
</div>
`;

exports[`GalleryScreen should render correctly with success fetch as favorites 1`] = `
<div
  className="sc-bczRLJ ljydhb"
>
  <div>
    VideoCard 
    videoId1
  </div>
  <div>
    VideoCard 
    videoId2
  </div>
  <div>
    VideoCard 
    videoId3
  </div>
</div>
`;

exports[`GalleryScreen should render correctly with success fetch as favorites with no results 1`] = `
<div
  className="sc-bczRLJ ljydhb"
>
  <p
    className="animate__animated animate__fadeIn"
  >
    You have no favorites yet
  </p>
</div>
`;

exports[`GalleryScreen should render correctly with success fetch as search 1`] = `
<div
  className="sc-bczRLJ ljydhb"
>
  <div>
    VideoCard 
    videoId1
  </div>
  <div>
    VideoCard 
    videoId2
  </div>
  <div>
    VideoCard 
    videoId3
  </div>
</div>
`;

exports[`GalleryScreen should render correctly with success fetch as search with no results 1`] = `
<div
  className="sc-bczRLJ ljydhb"
>
  <p
    className="animate__animated animate__fadeIn"
  >
    No results found for your search.
  </p>
</div>
`;
